src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs

```diff
@@ -720,37 +689,44 @@ private class LookUpNode
 
             // If a node is a project then it will hold these properties
             public NodeWarningProperties NodeWarningProperties { get; set; }
-
         }
 
-
         /// <summary>
-        /// A class to hold minimal version of project wide nowarn and package specific no warn for a project.
+        /// An immutable class to hold minimal version of project wide nowarn and package specific no warn for a project.
         /// </summary>
         public class NodeWarningProperties : IEquatable<NodeWarningProperties>
         {
+            // Empty NodeWarningProperties singleton - to avoid allocation for most-common case
+            internal static readonly NodeWarningProperties Empty = new NodeWarningProperties(null, null);
```

Comment:
I like this optimization so that projects with no suppressed warnings don't pay a penalty.  I'm curious if we could cache more.  For example, I've suppressed a warning or set of warnings in `Directory.Build.props` so every project has the exact same set.  Can we use the same object for each project in that case?  Essentially we'd only allocate a new instance if a project had a different set?  We could of course do this in a future change but wanted to throw it out there.  If we could quickly look up that an immutable instance has already been created that represents the warnings, we could just return that instead of allocating a new one.

Reply:
That's certainly an interesting idea!  I don't have a good approach in mind for caching the dictionary or set though - perhaps if we had the knowledge here about the warnings coming from a shared set.

Perhaps another way of handling that situation would be to add something to GetIntersect that checks if the two nodes are equal, and if so then return one.  Or, if just the hash sets or dictionaries are equal, just reuse it.  Right now it handles the ReferenceEqual scenario, but not value-equal.  Adding a value-equal check would optimize Memory at a cost to CPU.  But perhaps we could also fit this into the IsSubset check - perhaps that could have a special case for equals?

This could be an interesting thing to file a bug about - performance for common suppressed warning(s) scenario.

Reply:
I've filed https://github.com/NuGet/Home/issues/11878 to track this idea.